### ISSUE TITLE
Probable bug: Contextual inference and intellisense at call sites buggy in case of circular type parameter constraints like `T extends M<T>`

### ISSUE THREAD

**Issue by devanshj** [microsoft/TypeScript#44821]
# Bug Report

### 🔎 Search Terms

Circular type parameter constraint, intellisense, call site inference, contextual inference
Related - #44428 #40439

### 🕗 Version & Regression Information

Tested with TS 4.3.4

### ⏯ Playground Link

[Playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gWwFzwB4AVeEADwxBWDXgFliA+RgCj3iIEp4BeRjgFAYAngAcEDAMogIAMwA08ANK94AUTKQArqAIBrEMKSz40uYoBEei-AA+8CwHcLzXgPjwA3vCgB+fCha2ABGIDDu8IrB-vCBIWERimAxcaHhHorAKUFpifB6Md4AJPgqAL55joXwJQ5QNhUeFQKgkLAIyOhYRQCM+EwsGPhcvPzFw-AVrdBwiKiYNQBM-QBKrEPwK6QUVDReNfgAliiyYfAAGpPwvhfwASAAbmHcfJst4DMd890AzKuKJHIlGotBYWyBu1o43gx1OMFuZWut3uTxgnHWwxe-BWAgE2BYngiUHwPXkEWC+EWZI8enwvRYFnqnEUAHoWXNsKJDhAQGh8MJeUkkJyeRhDvMAkgIo46T0GUyPGyOVyeXz4AK0EKRSAxRLYlKypxcfjCR5ifBSeTKdT8nTFvKLMz4ErkJzubzJVrRKLxeh+bzpXaHdxnezXSqPerBcqfXqNQJDcaCUSSTaKfAqRFaTUfsHWWHhRG1SgkF7Y36o2hAzng4qC27Vf7NTGdb61fHDUA)

### 💻 Code

```ts
declare const m: <T extends M<T>>(m: T) => T
type M<Self, K = Exclude<keyof Self, "k" | "w">> =
  { a?: number
  , b?: number
  , c?: number
  , d?: number
  , k?: { $: K }
  , w?: { $: "a" }
  }

declare const $1: <T>(t: T) => { $: T }
declare const $2: <R>(t: R extends { $: infer X } ? X : never) => R
declare const $3: <R, T extends (R extends { $: infer X } ? X : never)>(t: T) => R

m({
  a: 1,
  b: 2,
  k: $1("a"), // compiles: yes, completions: no
  w: $1("a")  // compiles: yes, completions: no
})

m({
  a: 1,
  b: 2,
  k: $2("a"), // compiles: no, completions: yes
  w: $2("a")  // compiles: yes, completions: yes
})

m({
  a: 1,
  b: 2,
  k: $3("a"), // compiles: no, completions: yes
  w: $3("a")  // compiles: yes, completions: yes
})
```

### 🙁 Actual behavior

See code comments

### 🙂 Expected behavior

All three cases or at least `$2` and `$3` should compile and user should be able to see completions "a" and "b" for the first parameter of `$*` functions

It seems to be bug to me especially because in case of `k` for `$2` and `$3`, you do get the completions and even the tooltip shows correct type parameters, but somehow the return type isn't the one thats expected and it doesn't compile AND it works totally fine with `w`.

**Comment by devanshj** [comment]
[Workaround](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gWwFzwB4AVeEADwxBWDXgFliA+RgCj3iIEp4BeRjgFAYAngAcEDAMogIAMwA08ANK94AUTKQArqAIBrEMKSz40uYoBEei-AA+8CwHcLzXgPjwA3vCgB+fCha2ABGIDDu8IrB-vCBIWERimAxcaHhHorAKUFpifB6Md4AJPgqAL55joXwJQ5QNhUeFQKgkLAIyOhYRQAsuBEE8ABKeSTklNS0LEOkFFQ0XjX4AJYosmHwABrwZfC+W-ABIABuYZysER4Y+Fy8-DMAZIu1JM3YLJ4RUPgAjPIRwXwACZ-h49PheiwLPVOIoAPRwxBIbCiZYQEBofDCDFJZGidEYZaoTHwbFoCKOCE9KEwjwIpEotEYrE4hn4kCE4ks8llThAA). Although it doesn't work in my local more complex scenario, so will have to find another minimal repro where it fails.

```ts
declare const m: <T extends M<T>>(m: T) => T
type M<Self, K = Exclude<keyof Self, "k" | "w">> =
  { a?: number
  , b?: number
  , c?: number
  , d?: number
  , k?: { $: K }
  , w?: { $: "a" }
  }

declare const $4:
  < R
  , T extends (R extends { $: infer X } ? X : never)>(
    t: T) => R & { $: T }

m({
  a: 1,
  b: 2,
  k: $4("a"), // compiles: yes, completions: yes
  w: $4("a")  // compiles: yes, completions: yes
})
```

**Comment by RyanCavanaugh** [comment]
I think my comment in #44879 applies equally in this scenario -- `$3` is not a meaningful declaration.

It's a fairly normal occurrence for completions to show more options than might be strictly legal, since you might be "in the middle" of writing something and need to provide several completed items before the surrounding call becomes valid (the set of outcomes of which we have no ability to predict within a few hundred milliseconds).

`$2` seems to be expecting that inference will cause the result type `{ $: "a" }` to be produced, but this isn't an intended use of `infer`. It should really be written the way `$1` is.

**Comment by devanshj** [comment]
> It should really be written the way $1 is

Agreed (others are workarounds). I think it'd make that the user gets completions for the first parameter of `$1` because if you simply write what is being returned by `$1` [you do get the completions](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gWwFzwB4AVeEADwxBWDXgFliA+RgCj3iIEp4BeRjgFAYAngAcEDAMogIAMwA08ANK94AUTKQArqAIBrEMKSz40uYoBEei-AA+8CwHcLzXgPjwA3vCgB+fCha2ABGIDDu8IrB-vCBIWERimAxcaHhHorAKUFpifB6Md4AJPgqAL55joXwJQ5QNhUeFQLYLJ4RUPgAjPIRwfgATL0eevjF+Bb18GWKAPSziEjYogCWECBo+MIbSUui6xgrqJvw22gRjmM1E1NlHvOLy2sbWzuP+yCHx6-nZZxAA). Am I wrong in thinking this?

**Comment by RyanCavanaugh** [comment]
> ... I'd ideally expect that c gets inferred as `{ foo: number }` in second call too

If you want inference here, I'd remove the `NoInfer` wrapper around `C` in `assign`. It works as expected then -- `NoInfer` is doing its job.

**Comment by devanshj** [comment]
[Here's a version with no `NoInfer`s doesn't seem to work](https://www.typescriptlang.org/play?#code/MYJwpghgLmCyHABYEsB2YAUBvAUAAj2AHtUYAPKALjyzwDMijqBGPAXwBp88xSQBPargIEo-AA5hqAIjIBnKNDAA6CHLnIA5qmlcRPMmGDVgeALwA+PNnqMTyhkTwBqPKzYBKbmxyecOUEgYeCQ0TGFCEnIqGlsmN3Y9Hj5BPDUNbQxTS2taR3tHFwTPDzwAejLCPGQ5NNR+PAAjAFcoPDlEImaAGwATJrBY-LxUZoBbRrAQdmrUBUhe3y8cXqNuiHBIubbApRCUdGoAHgBhCwxVujQoZBJqfbDTi1KcrB8xSTwH9CfzGm5iKQwBRqCckrwoAIhNxRBIpO1IWhNEkCMCjNQslFgTETi8rM1UABrVBEADuqG8vn8q2A602gIUaXUWlQxzOGDoGMB0VBeLwuPMFm4tA+8NkCiUqmZ2l0BnR1m52N5gv57CAA)

**Comment by devanshj** [comment]
@RyanCavanaugh senpai could I get your word on this please? Because I'm still not sure what is the intended behaviour

**Comment by RyanCavanaugh** [comment]
@devanshj sorry, been snoozing on this notification because I knew it'd be interesting (usually implies time-consuming, but not this time). I would consider the linked repro there likely to be a bug; the contextual inference candidate for the `assign` call should not have fixed to `unknown` like that. Can you log a fresh issue with just that one?

**Comment by devanshj** [comment]
No problems [done](#44999). Though, Ryan, I'm asking this for the third time xD - I expect "a" and "b" to be suggested in completions, is that wrong?

Here's the most minimal repro... Afaict the assign issue and this are a consequence of the same bug.
```ts
declare const m: (x: { a: X, b: X }) => {}
type X = { $: "foo" | "bar" }
declare const $: <T>(x: T) => { $: T }

m({
  a: { $: "foo" },
  b: $("")
  //    ^|
  // expected completions: foo, bar
  // actual completions: none
})
```

**Comment by RyanCavanaugh** [comment]
I don't have a way to describe "expected" completion results; completions are either incorrect (nearly always a bug) or just "would be nicer if it happened more often". During an in-fact illegal call (which usually means inference has failed), we apply some heuristics in the language service to try to piece together what we think values that might make the call actually legal. If the inference needed is too complex, none of those heuristics will apply, and there won't be completions as a result. These are tricky to tweak but if you have a PR that fixes it, we'd probably merge it. I wouldn't consider this case to be a per se defect because inference has failed and without a working inference, we can only rely on the salvaging heuristics, which will always be imperfect.

**Comment by devanshj** [comment]
Thanks for the detailed explanation. Though I realized I missed an important nuance in my question. What I meant to ask was - Is it wrong to expect the type parameter to be inferred as "a" | "b" or in the above case "foo" | "bar"? (the completions would only be a part of that). Or in other words is the inference failure intended?

[It does work with `NoInfer` in the above case](https://www.typescriptlang.org/play?jsx=0&ts=4.4.0-beta#code/CYUwxgNghgTiAEYD2A7AzgF3gWwFzwAoAPfAb3inwA0AaeAI2vgF8BKeAXgD55TmAoDAE8ADgiqde8ACT4ARADMkSOfAA+8OfViqBoSLATJ0WWfAA8AFS7F8AOSQBJFApAwrXdtylnLLQaIIDs6u7taSANqWALpR8CBEGCAowGgUKELwAPzwAAzw+CggAG5u0fz82ASk-PAUZDLySiosNLUM+NIEcnKs7QD0-XV1AHpqA0MJYmBJwIhI2CIQIBgAlqho+M102jATFDMArlAQ84vLaxuFqCD8bEA), [but not in the original description issue for `k`](https://www.typescriptlang.org/play?jsx=0&ts=4.4.0-beta#code/CYUwxgNghgTiAEYD2A7AzgF3gWwFzwB4AVeEADwxBWDXgFliA+RgCj3iIEp4BeRjgFAYAngAcEDAMogIAMwA08ANK94AUTKQArqAIBrEMKSz40uYoBEei-AA+8CwHcLzXgPjwA3vCgB+fCha2ABGIDDu8IrB-vCBIWERimAxcaHhHorAKUFpifB6Md4AJPgqAL55joXwJQ5QNhUeFQKgkLAIyOhYtUwsGPgAckgAkiiyYUzcfF41+CQVIuLwQ6PjMEyqANpEALrbpBRUND4owvC+8AAM8AEgAG5hOwIC2CyeEVD4AIzyEcH4ACZfh49PgiiwLBZOMD4I4wRCoQIypwBEA). I've opened an [issue](https://github.com/microsoft/TypeScript/issues/45035) regarding this where I also sell it why it seems like a bug or if not a bug then sure is a good improvement becomes a feature request.
